File: 1987.cpp
#include<iostream>
#include<string>
#include<deque>
using namespace std;
deque<string> app;
int maxim=0;
char visit[26];
void searching(int cnt,int i,int j,int r,int c){
  visit[app[i][j]-65]=1;
  if(i>0&&visit[app[i-1][j]-65]==0){
    searching(cnt+1,i-1,j,r,c);
  }
  if(i<r-1&&visit[app[i+1][j]-65]==0){
    searching(cnt+1,i+1,j,r,c);
  }
  if(j>0&&visit[app[i][j-1]-65]==0){
    searching(cnt+1,i,j-1,r,c);
  }
  if(j<c-1&&visit[app[i][j+1]-65]==0){
    searching(cnt+1,i,j+1,r,c);
  }
  if(cnt>maxim)maxim=cnt;
  visit[app[i][j]-65]=0;
  return;
}
int main(void){
  int r,c;
  string str;
  cin>>r>>c;
  for(int i=0;i<r;i++){
    cin>>str;
    app.push_back(str);
  }
  searching(1,0,0,r,c);
  cout<<maxim<<endl;

  return 0;
}

File: 1987.py
input_str=list()
past=list()
def searching(cnt,i,j,r,c,maxim):
    past.append(input_str[i][j])
    if i>0 and past.count(input_str[i-1][j])==0:
        maxim=searching(cnt+1,i-1,j,r,c,maxim)
    if i<r-1 and past.count(input_str[i+1][j])==0:
        maxim=searching(cnt+1,i+1,j,r,c,maxim)
    if j>0 and past.count(input_str[i][j-1])==0:
        maxim=searching(cnt+1,i,j-1,r,c,maxim)
    if j<c-1 and past.count(input_str[i][j+1])==0:
        maxim=searching(cnt+1,i,j+1,r,c,maxim)
    past.pop()
    if maxim<cnt:
        maxim=cnt
    return maxim
r,c=input().split()
r=int(r)
c=int(c)
maxim=0
for i in range(r):
    tmp=input()
    input_str.append(tmp)
maxim=searching(1,0,0,r,c,maxim)
print(maxim)
